tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc

```diff
@@ -1371,18 +1371,45 @@ bool ReduceTransposer::IsReduceAxisSupported(
 Status ReduceTransposer::TransposeNode(TransposeContext* context,
                                        utils::MutableNodeView* node) {
   DCHECK(IsReduceOp(*node->node()));
-  if (!ShouldProcess(*context, *node) || !IsFaninPortRankN(*node, 0, 4) ||
+  const auto& regular_fanin = node->GetRegularFanin(0);
+  const auto* output_shape_attr =
+      regular_fanin.node_view()->GetAttr(kAttrOutputShape);
+  const auto& shape = output_shape_attr->list().shape(0);
+  const int rank = shape.dim_size();
+  std::string src_format = context->src_format;
+  std::string dst_format = context->dst_format;
+  // Update the format from 4D to 5D layout if necessary.
+  if (rank == 5) {
+    std::string src_format_3d = src_format == "NHWC" ? "NDHWC" : "NCDHW";
```

Comment:
This seems very specific to CPU to GPU in regards to data formats. Can we check if the original src_formats and dst_formats are NHWC and NCHW only before applying this?

Reply:
I think that is correct. The layout opt only does the NHWC<->NCHW transpose. You can find that here that the src/dst formats are predefined: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/grappler/optimizers/generic_layout_optimizer.cc#L38

Reply:
If I remembered correctly the actual transposers can support other layouts. The optimizer though restricts the layouts, along with some GPU specific bits. I think it is a bit safer to only allow the 5d upgrade in the transposers if the layouts are of NHWC and NCHW for now (if the optimizer is updated to support other layouts).

Reply:
Sure. I added one more barrier to restrict the optimizer only for NHWC/NCHW when 5d tensors are there.